### DISCUSSION TITLE
Trying to implement a way to change schema at runtime

### DISCUSSION THREAD

**Discussion by Datron** [diesel-rs/diesel#4267] [Q&A]
cargo version: `cargo 1.80.0 (376290515 2024-07-16)`
rust version: `rustc 1.80.1 (3f5fd8dd4 2024-08-06) (built from a source tarball)`

Hi Diesel Team,

I'm trying to build a way to change the table schema at runtime. This is to support a multi-tenant setup for a project I'm working on. Earlier I was setting the search path in the connection string and pooling these connections for a particular tenant, but as the number of tenants have increased; I have reached the max number of connections I can make to my postgres instance.

I tried following the advice in this [thread](https://github.com/diesel-rs/diesel/discussions/3865) and got an [implementation](https://github.com/juspay/diesel-multiconnection-manager/blob/runtime-schema-change/multi_connection_manager/src/pg_schema.rs) done, but I can't get it to work. I get the error:

```
the method `select` exists for struct `PgSchema<table>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`&PgSchema<table>: Table`
which is required by `&PgSchema<table>: QueryDsl`
`&mut PgSchema<table>: Table`
which is required by `&mut PgSchema<table>: QueryDsl`
```

Not sure what I am missing here. Am I on the right track? I'm looking for guidance and suggestions and how I can implement this correctly. 
[Example](https://github.com/juspay/diesel-multiconnection-manager/blob/runtime-schema-change/example/src/main.rs#L120)

**Comment by Datron** [comment]
Some info on diesel integration
version = "2.1.2"
features = ["postgres", "postgres_backend", "r2d2", "chrono", "uuid"]

**Comment by weiznich** [comment]
As listed in the documentation [`SelectDsl`](https://docs.diesel.rs/2.2.x/diesel/query_dsl/methods/trait.SelectDsl.html#implementors) is only implemented for 4 types. Your `PgSchema` type is neither one of the 4 concrete types nor does it implement `Table` as required by the fourth impl.

You have two options to fix that: Either implement `Table` or manually implement `SelectDsl`.

**Reply by Datron** [reply]
@weiznich I have implemented `SelectDsl` [here](https://github.com/juspay/diesel-multiconnection-manager/blob/runtime-schema-change/multi_connection_manager/src/pg_schema/dsl_impls.rs#L29) but it doesn't work, I get the same error.  How would I get this to work with Inserts as well? I don't see a `InsertDsl` trait.

I tried implementing the trait `Table` on my struct `PgSchema`, but this results in an issue where the compilation of my code never terminates.

**Reply by Datron** [reply]
If I import `diesel::query_dsl::select_dsl::SelectDsl` into the file and then try to select on a PgSchema type, I get
```
error[E0599]: the method `select` exists for struct `PgSchema<table>`, but its trait bounds were not satisfied
   --> example/src/main.rs:129:10
    |
128 |       let test_one_records = _users_schema
    |  ____________________________-
129 | |         .select(username)
    | |         -^^^^^^ method cannot be called on `PgSchema<table>` due to unsatisfied trait bounds
    | |_________|
    |
    |
    = note: the following trait bounds were not satisfied:
            `&PgSchema<table>: Table`
            which is required by `&PgSchema<table>: QueryDsl`
            `&PgSchema<table>: Table`
            which is required by `&PgSchema<table>: SelectDsl<_>`
            `&PgSchema<table>: Query`
            which is required by `&PgSchema<table>: SelectDsl<_>`
            `&PgSchema<table>: SelectDsl<_>`
            which is required by `&PgSchema<table>: SelectDsl<_>`
            `&mut PgSchema<table>: Table`
            which is required by `&mut PgSchema<table>: QueryDsl`
            `&mut PgSchema<table>: Table`
            which is required by `&mut PgSchema<table>: SelectDsl<_>`
            `&mut PgSchema<table>: Query`
            which is required by `&mut PgSchema<table>: SelectDsl<_>`
            `&mut PgSchema<table>: SelectDsl<_>`
            which is required by `&mut PgSchema<table>: SelectDsl<_>`

```

I'm wondering if I have to implement SelectDsl for all references to PgSchema or is there any other way?

**Reply by Datron** [reply]
Hey @weiznich , any ideas on what I could do here? I'm still stuck on this

**Reply by weiznich** [reply]
I've not forgotten about this, but I just don't have the capacity to provide an answer. After all this is still a free time project, so answering support question is something that needs to wait until I have time. If you need a more timely support channel you would need to think about a support contract.

**Reply by Datron** [reply]
Sure, thanks for your time. Just wanted to check in and see if you have any inputs.
